optaplanner-examples/src/main/java/org/optaplanner/examples/flightcrewscheduling/optional/score/FlightCrewSchedulingConstraintProvider.java

```diff
@@ -0,0 +1,63 @@
+package org.optaplanner.examples.flightcrewscheduling.optional.score;
+
+import java.util.Objects;
+
+import org.optaplanner.core.api.score.buildin.hardsoftlong.HardSoftLongScore;
+import org.optaplanner.core.api.score.stream.Constraint;
+import org.optaplanner.core.api.score.stream.ConstraintFactory;
+import org.optaplanner.core.api.score.stream.ConstraintProvider;
+import org.optaplanner.examples.flightcrewscheduling.domain.Employee;
+import org.optaplanner.examples.flightcrewscheduling.domain.FlightAssignment;
+
+public class FlightCrewSchedulingConstraintProvider implements ConstraintProvider {
+
+    @Override
+    public Constraint[] defineConstraints(ConstraintFactory constraintFactory) {
+        return new Constraint[]{
+                flightConflict(constraintFactory),
+                requiredSkill(constraintFactory),
+                employeeUnavailability(constraintFactory),
+                transferBetweenTwoFlights(constraintFactory),
+                firstAssignmentDepartingFromHome(constraintFactory),
+                lastAssignmentArrivingAtHome(constraintFactory)
+        };
+    }
+
+    private Constraint flightConflict(ConstraintFactory constraintFactory) {
+        return constraintFactory.fromUniquePair(FlightAssignment.class)
+                .filter((first, second) -> Objects.equals(first.getEmployee(), second.getEmployee()))
```

Comment:
This should use a joiner, for performance:

constraintFactory.fromUniquePair(FlightAssignment.class, equals(FlightAssignment::getEmployee))

Reply:
see also javadoc of join:

     * Important: {@link BiConstraintStream#filter(BiPredicate) Filtering} this is slower and less scalable
     * than a {@link #join(UniConstraintStream, BiJoiner)},
     * because it doesn't apply hashing and/or indexing on the properties,
     * so it creates and checks every combination of A and B.